sdk/src/main/scala/com/horizen/account/api/rpc/service/EthService.scala

```diff
@@ -348,32 +386,123 @@ class EthService(val scNodeViewHolderRef: ActorRef, val nvtimeout: FiniteDuratio
+    val currentBlockNumber = Numeric.cleanHexPrefix(blockNumber)
+    val previousBlockNumber = Numeric.cleanHexPrefix((java.lang.Long.decode(currentBlockNumber) - 1).toHexString)
+
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, previousBlockNumber) { tagStateView =>
+        {
+          val requestedBlockId = getBlockIdByTag(nodeView, currentBlockNumber)
+          val requestedBlock = nodeView.history.getBlockById(requestedBlockId).get()
+          val transactions = requestedBlock.transactions
+
+          for (mcBlockRefData <- requestedBlock.mainchainBlockReferencesData) {
+            tagStateView.applyMainchainBlockReferenceData(mcBlockRefData).get
+          }
+
+          val evmResults = transactions.map(tx => {
+            Evm.Trace(
+              tagStateView.getStateDbHandle,
+              if (tx.getFrom == null) null
+              else tx.getFrom.bytes(),
+              if (tx.getTo == null) null
+              else tx.getTo.bytes(),
+              tx.getValue,
+              tx.getData,
+              tx.getGasLimit,
+              tx.getGasPrice,
+              null
+            )
+          })
+
+          new DebugTraceBlockByIdView(evmResults.toArray)
+        }
+      }
+    }
+  }
+
+  @RpcMethod("debug_traceTransaction")
+  def traceTransaction(transactionHash: String): DebugTraceTransactionView = {
+    val requestedTransaction = getTransactionAndReceipt(transactionHash) { (tx, receipt) =>
+      new EthereumTransactionView(receipt, tx)
+    }.orNull
+
+    val currentBlockId = Numeric.cleanHexPrefix(requestedTransaction.getBlockHash)
+    val currentBlockNumber = requestedTransaction.getBlockNumber
+    val previousBlockNumber = Numeric.cleanHexPrefix((java.lang.Long.decode(currentBlockNumber) - 1).toHexString)
+
+    val evmResult: EvmResult = new EvmResult()
+
+    applyOnAccountView { nodeView =>
+      getStateViewAtTag(nodeView, previousBlockNumber) { tagStateView =>
+        {
+          val currentBlock = nodeView.history.getBlockById(currentBlockId).get()
+          val gasPool = new GasPool(BigInteger.valueOf(currentBlock.header.gasLimit))
+          val transactions = currentBlock.transactions
+
+          for (mcBlockRefData <- currentBlock.mainchainBlockReferencesData) {
+            tagStateView.applyMainchainBlockReferenceData(mcBlockRefData).get
+          }
+
+          breakable {
+            for ((tx, i) <- transactions.zipWithIndex) {
+              if (tx.id == Numeric.cleanHexPrefix(transactionHash)) {
+                val txResult = Evm
+                  .Trace(
+                    tagStateView.getStateDbHandle,
+                    if (tx.getFrom == null) null
+                    else tx.getFrom.bytes(),
+                    if (tx.getTo == null) null
+                    else tx.getTo.bytes(),
+                    tx.getValue,
+                    tx.getData,
+                    tx.getGasLimit,
+                    tx.getGasPrice,
+                    null
```

Comment:
why we keep `null` context?
We may specify the proper one as it is in the state. See the latest changes by JC.